cowprotocol/cowswap
# feat(tokens): migrate user added data from legacy state

# Summary

During the refactoring, some data was moved from redux to jotai atoms.

1. User-added tokens: `redux_localstorage_simple_user.tokens` -> `userAddedTokensAtom:v1`
2. User-added lists: `redux_localstorage_simple_lists` -> `allTokenListsInfoAtom:v2`
3. Also, favorite tokens were moved from `redux_localstorage_simple_user.favouriteTokens` to `favouriteTokensAtom`. I didn't migrate them because it looks like it doesn't worth it.
4. `redux_localstorage_simple_user.pairs` was just removed, now `tokensStateAtom` performs its functions. 

  # To Test

1. Open https://dev.swap.cow.fi/
2. Add some custom tokens and lists
3. Keep some newly added lists enabled and some disabled
4. Run `downloadLocalStorage()` in the console (see the snipped below). It will download a file with the localStorage state
5. Open Dapp from this PR
6. Clean all browser data for this URL: Console -> Application -> Storage -> "Clear site data"
7. Copy the content of the file from step 4 and run in the console: `var state = <PUT_THE_FILE_CONTENT>`, press Enter
8. Run the code in the console: `localStorage.clear(); var stateObj = JSON.parse(state); Object.keys(stateObj).forEach(key => localStorage.setItem(key, stateObj[key]))`. It will apply the localStorage state from dev.swap.cow.fi into the current tab
9. Reload the page
- [ ] TOKENS_REFACTORING_MIGRATED should be added in the localStorage
- [ ] User-added tokens should be the same as in dev.swap.cow.fi
- [ ] User-added lists should be the same as in dev.swap.cow.fi including enabled/disabled state
- [ ] Favourite tokens should be reset to defaults
- [ ] `redux_localstorage_simple_lists` should be removed from localStorage
- [ ] `redux_localstorage_simple_user.favouriteTokens` should be removed from localStorage
- [ ] `redux_localstorage_simple_user.tokens` should be removed from localStorage
- [ ] `redux_localstorage_simple_user.pairs` should be removed from localStorage
10. Open the PR in incognito mode
- [ ] there should not be user-added tokens
- [ ] there should not be user-added lists

### Download localStorage snippet:
```
function downloadLocalStorage() {
    const storageObj = JSON.stringify(Object.keys(localStorage).reduce((acc, key) => {
        acc[key] = localStorage.getItem(key)

        return acc
    }, {}));
    
    const a = document.createElement('a');
    a.style.display = 'none';
    a.href = "data:text/json;charset=utf-8," + encodeURIComponent(JSON.stringify(storageObj))
    // the filename you want
    a.download = 'todo-1.json';
    document.body.appendChild(a);
    a.click();
}
```
